File: backend/agroforestry_systems_design/model/canteiro.py
from sqlalchemy import Column, String, Integer, JSON

from  model import Base

class Canteiro(Base):
    __tablename__ = 'canteiros'

    id_canteiro = Column(Integer, primary_key=True)
    nome_canteiro = Column(String(140), unique=True, nullable=False)
    x_canteiro = Column(Integer, nullable=False)
    y_canteiro = Column(Integer, nullable=False)
    plantas_canteiro = Column(JSON(100000), nullable=False)

    def __init__(self, nome_canteiro:str, x_canteiro:int, y_canteiro:int, plantas_canteiro:dict):
        """

        Cria um Canteiro

        Parametros:
            nome_canteiro: O nome da canteiro
            x_canteiro: Valor do eixo X do canteiro
            y_canteiro: Valor do exico Y do canteiro
            plantas_canteiro: Plantas de cada estrato
        """
        self.nome_canteiro = nome_canteiro
        self.x_canteiro = x_canteiro
        self.y_canteiro = y_canteiro
        self.plantas_canteiro = plantas_canteiro
        
    def distribuir_plantas(self):
        
        # Definindo o canteiro
        canteiro = {
            "emergente": [],
            "alto": [],
            "medio": [],
            "baixo": []
        }
    
        canteiro_x = int(self.x_canteiro)
        canteiro_y = int(self.y_canteiro)
        
        # Ordenando as plantas por tempo de colheita (priorizando colheita mais rápida)
        plantas_ordenadas = sorted(self.plantas_canteiro['plantas'], key=lambda x: int(x['tempo_colheita']))
    
        # Distribuindo as plantas nos estratos
        for planta in plantas_ordenadas:
            sombra = int(planta['sombra'])
            espacamento = int(planta['espacamento'])
            estrato = planta['estrato']
            
            print('planta: ', planta)
    
            # Calculando a área disponível
            area_disponivel = canteiro_x * canteiro_y
            # Calculando a área que a planta pode ocupar no estrato
            estrato_disponivel = area_disponivel * (sombra / 100)
            # Calculando a área que a planta pode ocupar
            area_planta = (espacamento ** 2)
    
            # Verifica se a planta cabe no espaço disponível
            if area_planta > estrato_disponivel:
                continue
            num_plantas_possiveis = int(estrato_disponivel // area_planta)
            if num_plantas_possiveis == 0:
                continue
                
            num_plantas_y = 1
            num_plantas_x = num_plantas_possiveis
            espacamento_x = canteiro_x // (num_plantas_x + 1)
            espacamento_y = canteiro_y // (num_plantas_y + 1)
                
            while espacamento_x < espacamento:
                num_plantas_x = num_plantas_possiveis
                if espacamento_y < espacamento:
                    continue
                num_plantas_y += 1
                num_plantas_x /= num_plantas_y
                espacamento_x = int(canteiro_x // (num_plantas_x + 1))
                espacamento_y = int(canteiro_y // (num_plantas_y + 1))
            num_plantas_x = int(num_plantas_x)
            num_plantas_y = int(num_plantas_y)
            y = espacamento_y
    
            for i in range(num_plantas_y):
                x = espacamento_x
                
                for j in range(num_plantas_x):
                    
                    canteiro[estrato].append({
                        "nome_planta": planta['nome_planta'],
                        "estrato": estrato,
                        "posicao": [x, y],
                        "diametro": espacamento,
                        "tempo_colheita": planta['tempo_colheita']
                    })
                    x += espacamento_x
                y += espacamento_y
    
        self.plantas_canteiro_destribuidas = canteiro
        
    def __repr__(self):
        return f'Canteiro("{self.nome_canteiro}","{self.plantas_canteiro}")'

File: backend/agroforestry_systems_design/src/models/canteiro.js
import { DataTypes } from 'sequelize';
import sequelize from '../config/db.js';

const Canteiro = sequelize.define('Canteiro', {
  id_canteiro: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  nome_canteiro: {
    type: DataTypes.STRING,
    allowNull: false,
    unique: true
  },
  x_canteiro: DataTypes.INTEGER,
  y_canteiro: DataTypes.INTEGER,
  plantas_canteiro: {
    type: DataTypes.JSON, // Armazena array de plantas
    allowNull: false
  }
}, {
  tableName: 'canteiro',
  timestamps: false
});

Canteiro.prototype.distribuirPlantas = function () {
  // Definindo o canteiro por estrato
  const canteiroDistribuido = {
    emergente: [],
    alto: [],
    medio: [],
    baixo: []
  };

  const canteiro_x = parseInt(this.x_canteiro, 10);
  const canteiro_y = parseInt(this.y_canteiro, 10);

  // Ordena as plantas por tempo de colheita (priorizando colheita mais rápida)
  const plantas_ordenadas = [...(this.plantas_canteiro.plantas || [])].sort(
    (a, b) => parseInt(a.tempo_colheita, 10) - parseInt(b.tempo_colheita, 10)
  );

  // Distribui as plantas nos estratos
  for (const planta of plantas_ordenadas) {
    const sombra = parseInt(planta.sombra, 10);
    const espacamento = parseInt(planta.espacamento, 10);
    const estrato = planta.estrato;

    // Calcula área disponível
    const area_disponivel = canteiro_x * canteiro_y;
    const estrato_disponivel = area_disponivel * (sombra / 100);
    const area_planta = espacamento ** 2;

    // Verifica se a planta cabe no espaço disponível
    if (area_planta > estrato_disponivel) continue;
    let num_plantas_possiveis = Math.floor(estrato_disponivel / area_planta);
    if (num_plantas_possiveis === 0) continue;

    let num_plantas_y = 1;
    let num_plantas_x = num_plantas_possiveis;
    let espacamento_x = Math.floor(canteiro_x / (num_plantas_x + 1));
    let espacamento_y = Math.floor(canteiro_y / (num_plantas_y + 1));

    while (espacamento_x < espacamento) {
      num_plantas_x = num_plantas_possiveis;
      if (espacamento_y < espacamento) continue;
      num_plantas_y += 1;
      num_plantas_x /= num_plantas_y;
      espacamento_x = Math.floor(canteiro_x / (num_plantas_x + 1));
      espacamento_y = Math.floor(canteiro_y / (num_plantas_y + 1));
    }
    num_plantas_x = Math.floor(num_plantas_x);
    num_plantas_y = Math.floor(num_plantas_y);
    let y = espacamento_y;

    for (let i = 0; i < num_plantas_y; i++) {
      let x = espacamento_x;
      for (let j = 0; j < num_plantas_x; j++) {
        canteiroDistribuido[estrato].push({
          nome_planta: planta.nome_planta,
          estrato: estrato,
          posicao: [x, y],
          diametro: espacamento,
          tempo_colheita: planta.tempo_colheita
        });
        x += espacamento_x;
      }
      y += espacamento_y;
    }
  }

  this.plantas_canteiro_destribuidas = canteiroDistribuido;
  return canteiroDistribuido;
};

export default Canteiro;